enhancements/ingress/client-tls.md

```diff
@@ -0,0 +1,325 @@
+
+	// allowedSubjectPatterns specifies a list of regular expressions that
+	// should be matched against the distinguished name on a client
+	// certificate to filter requests.  If this list is empty, no filtering
+	// is performed.  If the list is nonempty, then at least one pattern
+	// must match a client certificate's distinguished name or else the
+	// ingress controller rejects the certificate and denies the connection.
+	//
+	// +optional
+	AllowedSubjectPatterns []string `json:"allowedSubjectPatterns,omitempty"`
+}
+```
+
+The `ClientCertificatePolicy` type accepts either one of two values: `Required`
+or `Optional`:
+
+```go
+// ClientCertificatePolicy describes the policy for client certificates.
+// +kubebuilder:validation:Enum="";Required;Optional
+type ClientCertificatePolicy string
+
+const (
+	// ClientCertificatePolicyRequired indicates that a client certificate
+	// should be required.
+	ClientCertificatePolicyRequired ClientCertificatePolicy = "Required"
+
+	// ClientCertificatePolicyOptional indicates that a client certificate
+	// should be requested but not required.
+	ClientCertificatePolicyOptional ClientCertificatePolicy = "Optional"
+)
+```
+
+If the user does not specify `spec.clientTLS`, then client TLS is not enabled,
+which means that the IngressController does not request client certificates on
+TLS connections.  If `spec.clientTLS` is specified, then the IngressController
+does request client certificates, and `spec.clientTLS.clientCertificatePolicy`
+must be specified to indicate whether the IngressController should reject
+clients that do not provide valid certificates.
+
+The optional `ClientCA` field may be used to specify a reference to a configmap
+that is in the same namespace as the IngressController and contains a CA
+certificate bundle.  If `ClientCA` is omitted, then the IngressController uses
+the system trust bundle.  If `ClientCA` is specified, then the IngressController
+uses the provided bundle to verify client certificates, and the
+IngressController does not start serving unless and until it can read the CA
+certificate bundle as well as the certificate revocation list (CRL) for each
+certificate therein that specifies one or more destination points for retrieving
+that certificate's CRL.
+
+Finally, the optional `AllowedSubjectPatterns` field may be used to specify a
+list of patterns.  If the field is specified, then the IngressController rejects
+any client certificate that does not match at least one of the provided
+patterns.
+
+The following IngressController enables mandatory use of client certificates,
+verified using the CA certificates in the `router-ca-certs-default` configmap,
+and requiring that client certificates match the provided subject:
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  clientTLS:
+    clientCertificatePolicy: Required
```

Comment:
Consider adding an optional `crlDistributionPoints` field where you can override those found in the CA. This would help in the scenarios where the distribution point is down or changed, or if the CA does not have the extensions.

Reply:
Would it be sufficient to allow the user to specify the CRLs (for example, using a PVC, as you suggested in another comment) rather than the distribution points?